functional/kata-monitor/run.sh

```diff
@@ -0,0 +1,282 @@
+# - a CRI container manager (container engine) will be up and running
+# - crictl is installed and configured
+# - the kata-monitor binary is available on the host
+#
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+[ -n "${BASH_VERSION:-}" ] && set -o errtrace
+[ -n "${DEBUG:-}" ] && set -o xtrace
+
+readonly MONITOR_HTTP_ENDPOINT="127.0.0.1:8090"
+# we should collect few hundred metrics, let's put a reasonable minimum
+readonly MONITOR_MIN_METRICS_NUM=200
+CRI_RUNTIME=${CRI_RUNTIME:-"crio"}
+RUNTIME=${RUNTIME:-"containerd-shim-kata-v2"}
+KATA_MONITOR_BIN=${KATA_MONITOR_BIN:-$(command -v kata-monitor)}
+IAM=$(whoami)
+TMPATH=$(mktemp -d -t kata-monitor-test-XXXXXXXXX)
+METRICS_FILE="${TMPATH}/metrics.txt"
+MONITOR_LOG_FILE="${TMPATH}/kata-monitor.log"
+CACHE_UPD_TIMEOUT_SEC=${CACHE_UPD_TIMEOUT_SEC:-20}
+POD_ID=""
+CID=""
+RUNC_POD_ID=""
+RUNC_CID=""
+CURRENT_TASK=""
+
+FALSE=1
+TRUE=0
+
+trap error_with_msg ERR
+
+title() {
+	local step="$1"
+	echo -e "\n* STEP: $step"
+}
+
+echo_ok() {
+	local msg="$1"
+
+	echo "OK: $msg"
+}
+
+# quiet crictrl
+qcrictl() {
+	crictl "$@" > /dev/null
+}
+
+# this is just an hash of current date (+ nanoseconds)
+gen_unique_id() {
+	date +%T:%N | md5sum | cut -d ' ' -f 1
+}
+
+error_with_msg() {
+	local msg=${1:-"cannot $CURRENT_TASK"}
+
+	trap - ERR
+	echo -e "\nERROR: $msg"
+	echo -e "\nkata-monitor logs:\n----------------"
+	cat "$MONITOR_LOG_FILE"
+	echo -e "\nkata-monitor testing: FAILED!"
+	cleanup
+	exit 1
+}
+
+cleanup() {
+	stop_workload
+	stop_workload "$RUNC_CID" "$RUNC_POD_ID"
+
+	[ -d "/proc/$KATA_MONITOR_PID" ] \
+		&& kill -9 "$KATA_MONITOR_PID"
+
+	rm -rf "$TMPATH"
+}
+
+create_sandbox_json() {
+	local uid_name_suffix="$(gen_unique_id)"
+	local sbfile="$TMPATH/sandbox-$uid_name_suffix.json"
+
+	cat <<OEM >$sbfile
+{
+	"metadata": {
+		"name": "nginx-$uid_name_suffix",
+		"namespace": "default",
+		"attempt": 1
+	},
+	"logDirectory": "/tmp",
+	"linux": {
+	}
+}
+OEM
+	echo "$sbfile"
+}
+
+create_container_json() {
+	local uid_name_suffix="$(gen_unique_id)"
+	local cntfile="$TMPATH/container-$uid_name_suffix.json"
+
+	cat <<OEM >$cntfile
+{
+	"metadata": {
+		"name": "busybox"
+	},
+	"image":{
+		"image": "busybox"
+	},
+	"command": [
+		"top"
+	],
+	"log_path":"busybox.log",
+	"linux": {
+	}
+}
+OEM
+	echo "$cntfile"
+}
+
+start_workload() {
+	local runtime=${1:-}
+	local args=""
+	local sbfile=""
+	local cntfile=""
+
+	[ -n "$runtime" ] && args="-r $runtime"
+
+	sbfile="$(create_sandbox_json)"
+	cntfile="$(create_container_json)"
+
+	POD_ID=$(crictl runp $args $sbfile)
+	CID=$(crictl create $POD_ID $cntfile $sbfile)
+	qcrictl start $CID
+}
+
+stop_workload() {
+	local cid="${1:-$CID}"
+	local pod_id="${2:-$POD_ID}"
+	local check
+
+	check=$(crictl pods -q -id $pod_id)
+	[ -z "$check" ] && return
+
+	qcrictl stop $cid
+	qcrictl rm $cid
+
+	qcrictl stopp $pod_id
+	qcrictl rmp $pod_id
+}
+
+is_sandbox_there() {
+	local podid=${1}
+	local sbs s
+
+	sbs=$(curl -s ${MONITOR_HTTP_ENDPOINT}/sandboxes)
```

Comment:
@fgiudici, @liubin - I cannot get `/sandboxes` to work - it doesn't generate an HTTP error, but also doesn't show any output even when a sb is running?!?

Reply:
@liubin - thanks for checking. I see the problem - if you run it _after_ a container is started, you get nothing. I've raised https://github.com/kata-containers/kata-containers/issues/3704 as we should add some docs for this tool.

Reply:
> @liubin - thanks for checking. I see the problem - if you run it _after_ a container is started, you get nothing.

Well, actually when kata-monitor starts it does collect already running kata sandboxes, so it should work.
What may happen is that if you start kata-monitor and no kata sbs have ever been started from the host boot, the /run/vc/sbs dir will not exist - so kata-monitor will not be able to watch the dir.
In this particular case, kata-monitor will poll every 60 seconds to see if the directory has been created, so if you start your very first ever kata pod __after__ kata-monitor has been started, you may get a delay up to 60 seconds to have it reported by kata-monitor.

> I've raised [kata-containers/kata-containers#3704](https://github.com/kata-containers/kata-containers/issues/3704) as we should add some docs for this tool.

Yep, thank you, since I have done few changes I think it would be fair if I take care of that. I would like anyway to do it after  https://github.com/kata-containers/kata-containers/pull/3553 gets merged.

Reply:
Thanks @fgiudici ! :smile:

Reply:
@fgiudici - I'm using containerd and even though I have 3 sbs's in `/run/vc/sbs/`, starting `kata-monitor` in trace mode shows zero sandboxes in its logs and nothing is returned from `curl -s localhost:8090/sandboxes`. It looks like it's not scanning for existing directory entries on startup which would be a bug I guess.

Reply:
@jodh-intel, I would try to run:
`kata-monitor --log-level trace`

Reply:
Yep - I've got the output of that on https://github.com/kata-containers/kata-containers/issues/3705.